### PR TITLE
Nuget packages!

### PR DESCRIPTION
**Pull request by diddledan**
Reference #37 
This PR has many changes which are all aimed at nuget packages. I am able to successfully build locally and create .nupkg files which I am then able to install into a new ASP.NET application in Visual Studio. The resultant project is then able to successfully run PHP code. All that is required once the nupkg files are created is to point nuget at the files and install the core, core.parsers, core.codedom and classlibrary packages. Everything else is optional. Ideally we'll be able to create a metapackage which can pull-in the essentials in-one-go, but that's for the future.

### CODE REVIEW

**Comment by diddledan** [comment]
oh, hang-on, I appear to have missed-off the web.config automatic updating.

**Comment by diddledan** [comment]
that's better.

**Comment by diddledan** [comment]
the changes between my previous comment and this are a rebased version of the original commits above.

File: Source/ClassLibrary/ClassLibrary.csproj [hunk 1]
```diff
@@ -36,17 +36,33 @@
...
     <RegisterForComInterop>false</RegisterForComInterop>
     <RemoveIntegerChecks>false</RemoveIntegerChecks>
     <CodeAnalysisRuleSet>AllRules.ruleset</CodeAnalysisRuleSet>
+    <DebugType>pdbonly</DebugType>
+  </PropertyGroup>
+  <PropertyGroup Condition="'$(Configuration)|$(Platform)' == 'ReleaseGAC|AnyCPU'">
+    <OutputPath>bin\ReleaseGAC\</OutputPath>
+    <DefineConstants>TRACE</DefineConstants>
+    <BaseAddress>4194304</BaseAddress>
+    <DocumentationFile>C:\Users\yabea\Documents\GitHubVisualStudio\Phalanger\Source\Build\..\..\Deployment\Release\PhpNetClassLibrary.xml</DocumentationFile>
```
Comment (weirdan) [inline Source/ClassLibrary/ClassLibrary.csproj]: Wouldn't this cause build to fail for everyone not named 'yabea', let alone builds on other operating systems (e.g. Linux/mono)?

File: Source/Extensions/SQLite/Extension.SQLite.csproj [hunk 2]
```diff
@@ -89,11 +96,27 @@
...
   <Import Project="..\..\Build\Common.Build.targets" />
-  <!-- To modify your build process, add your task inside one of the targets below and uncomment it. 
+  <Import Project="..\..\..\Solutions\packages\System.Data.SQLite.Core.1.0.98.1\build\net40\System.Data.SQLite.Core.targets" Condition="Exists('..\..\..\Solutions\packages\System.Data.SQLite.Core.1.0.98.1\build\net40\System.Data.SQLite.Core.targets')" />
+  <Target Name="EnsureNuGetPackageBuildImports" BeforeTargets="PrepareForBuild">
+    <PropertyGroup>
+      <ErrorText>This project references NuGet package(s) that are missing on this computer. Use NuGet Package Restore to download them.  For more information, see http://go.microsoft.com/fwlink/?LinkID=322105. The missing file is {0}.</ErrorText>
+    </PropertyGroup>
+    <Error Condition="!Exists('..\..\..\Solutions\packages\System.Data.SQLite.Core.1.0.98.1\build\net40\System.Data.SQLite.Core.targets')" Text="$([System.String]::Format('$(ErrorText)', '..\..\..\Solutions\packages\System.Data.SQLite.Core.1.0.98.1\build\net40\System.Data.SQLite.Core.targets'))" />
+  </Target>
+<<<<<<< HEAD
```
Comment (weirdan) [inline Source/Extensions/SQLite/Extension.SQLite.csproj]: Here you have unresolved conflict (XML parser will likely choke on it).

**Comment by diddledan** [comment]
both the issues highlighted by @weirdan are now fixed.

**Comment by weirdan** [comment]
Could you also fix this?:
```
CSC: error CS1904: `CS0618' is not a valid warning number
```
It's in Source/Extensions/GetText/Extensions.GetText.csproj file.

**Comment by weirdan** [comment]
Also, is it possible to `nuget pack` with `.nuspec` files instead of `.csproj`? Nuget-mono doesn't seem to support packing with project files: http://stackoverflow.com/a/34394716/450611

Another improvement I would suggest is automatic package restore for non-VS environments. It's rather easy, just add another target and make BeforeBuild and BeforeRebuild depend on it 
```xml
  <Target Name="RestorePackages">
         <Exec Command="&quot;nuget&quot; restore &quot;$(MSBuildThisFileDirectory)..\..\Solutions\Phalanger.sln&quot;"/>
   </Target>
```

**Comment by diddledan** [comment]
with regards to building nuget packages under mono, this is known to fail for various reasons, including the one you cite, but also e.g. nuget/home#913

**Comment by diddledan** [comment]
I've uploaded example builds of nuget packaging at http://phalangernuget.azurewebsites.net/

File: Source/ClassLibrary/ClassLibrary.csproj [hunk 3]
```diff
@@ -36,23 +36,39 @@
...
   <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Release|AnyCPU' ">
     <AllowUnsafeBlocks>false</AllowUnsafeBlocks>
-    <BaseAddress>419430400</BaseAddress>
+    <BaseAddress>4194304</BaseAddress>
     <CheckForOverflowUnderflow>false</CheckForOverflowUnderflow>
     <DefineConstants>TRACE</DefineConstants>
     <DocumentationFile>Bin\Release\PhpNetClassLibrary.xml</DocumentationFile>
-    <DebugSymbols>false</DebugSymbols>
-    <FileAlignment>4096</FileAlignment>
+    <DebugSymbols>true</DebugSymbols>
```
Comment (jakubmisek) [inline Source/ClassLibrary/ClassLibrary.csproj]: set it to false in Release

**Comment by jakubmisek** [comment]
Thats great! One question - how is the versioning handled?

**Comment by diddledan** [comment]
at the moment it's done manually via editing the nuspec. We could possibly introduce a change to, or script in addition to, the file `Tools/UpdateFileVersionFromChangeset.cmd` (and .sh equivalent)

**Comment by jakubmisek** [comment]
Versioning is pretty important here, at least it must be in sync with https://github.com/DEVSENSE/Phalanger/blob/master/Source/Build/AssemblyInfoCommon.cs#L20 and the minor version below ... is it possible to make it somehow ? so the nuget package would have the same version

**Comment by diddledan** [comment]
The commit 6974287 tells nuget to use the details in each project's metadata after they're compiled rather than being manually altered. This means that version numbers should be synchronised now as requested.

920e736 fixes nuget's building of packages complaining that version numbers were unspecified for projects depending on other projects in the solution (referring to ourselves).

2a16875 can be split-out to a separate merge if needed and is a simplification and unification of all the msbuild settings and tasks set in each .csproj file moving them all into the common build tasks and settings files where possible.

**Comment by diddledan** [comment]
about version numbering. nuget uses SEMVER numbers which are different to the original .net method (three numbers instead of four). There's [a discussion piece about the implications of nuget/semver with .net libraries](https://codingforsmarties.wordpress.com/2016/01/21/how-to-version-assemblies-destined-for-nuget/).

**Pull request merged**